capstone_project/src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -28,52 +38,61 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
 
-  private void addUWBToSchools() {
-    ArrayList<String> uwbNews =
-        new ArrayList<String>(
-            Arrays.asList(
-                "https://news.uci.edu/2020/06/30/virtual-nurses-make-a-real-difference/",
-                "https://news.uci.edu/2020/07/07/uci-chancellor-emeritus-michael-v-drake-named-university-of-california-president/",
-                "https://news.uci.edu/2020/06/18/uci-podcast-jessica-millward-on-the-meaning-and-importance-of-juneteenth/"));
-    School uwb = new School("UWB", 47.758821, -122.190725, uwbNews);
-    schools.add(uwb);
-  }
+    for (Entity entity : results.asIterable()) {
+      String schoolName = (String) entity.getProperty("name");
+      double schoolLatitude = (double) entity.getProperty("latitude");
+      double schoolLongitude = (double) entity.getProperty("longitude");
+      School retrievedSchool = new School(schoolName, schoolLatitude, schoolLongitude);
+      schoolsSet.add(retrievedSchool);
+    }
 
-  /*Add some hard-coded School objects to the array of schools that will later be converted into markers on the map.*/
-  public MapServlet() {
-    addUCIToSchools();
-    addUWToSchools();
-    addUWBToSchools();
+    // Send the School objects retrieved from Datastore.
+    response.setContentType("application/json;");
+    Gson gson = new Gson();
+    response.getWriter().println(gson.toJson(schoolsSet));
   }
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("application/json;");
-    Gson gson = new Gson();
-    response.getWriter().println(gson.toJson(schools));
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Initialize datastore object.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get the input from the school entry form.
+    String schoolName = getParameter(request, "name-input");
+    double schoolLatitude = Double.parseDouble(getParameter(request, "latitude-input"));
+    double schoolLongitude = Double.parseDouble(getParameter(request, "longitude-input"));
+
+    // Check if the School is already in Datastore.
+    Filter nameFilter = new FilterPredicate("name", FilterOperator.EQUAL, schoolName);
+    Query query = new Query("School").setFilter(nameFilter);
+    PreparedQuery results = datastore.prepare(query);
+
+    // If it is not, create a schoolEntity for the School and add it to Datastore.
+    if (results.countEntities(FetchOptions.Builder.withDefaults()) == 0) {
+      Entity schoolEntity = new Entity("School");
+      schoolEntity.setProperty("name", schoolName);
+      schoolEntity.setProperty("latitude", schoolLatitude);
+      schoolEntity.setProperty("longitude", schoolLongitude);
+      datastore.put(schoolEntity);
+    }
```

Comment:
I'm slightly leaning towards you should create a test for this section just to ensure it does work. I'm assuming a test could be that datastore.put is called if the school doesn't exist and another test could check that datastore.put is not called when the school already exists.

Reply:
@Leoism just added these!